File: src/providers/redpill.ts
import axios from "axios";
import type { AttestationReport, ServiceProvider } from "../types";

export class RedpillProvider implements ServiceProvider {
	name = "redpill";
	private apiBase = "https://api.redpill.ai/v1";

	async fetchReport(modelId: string): Promise<AttestationReport> {
		const reportUrl = `${this.apiBase}/attestation/report?model=${encodeURIComponent(modelId)}`;
		console.log(`[Redpill] Fetching from ${reportUrl}`);

		const response = await axios.get(reportUrl);
		const data = response.data;

		if (!data.intel_quote) {
			throw new Error("Redpill report missing intel_quote");
		}

		return {
			intel_quote: data.intel_quote,
			nvidia_payload: data.nvidia_payload,
			raw: data,
		};
	}

	async listModels(): Promise<string[]> {
		console.log(`[Redpill] Fetching models from ${this.apiBase}/models`);
		const response = await axios.get(`${this.apiBase}/models`);
		// Handle potential response wrappers nicely
		const data = response.data;
		const models = Array.isArray(data) ? data : data.data || [];

		// biome-ignore lint/suspicious/noExplicitAny: API response is dynamic
		return models.map((m: any) => m.id);
	}
}

File: confidential_verifier/providers/redpill.py
import requests
import secrets
import json
from typing import List
from .base import ServiceProvider
from ..types import AttestationReport


class RedpillProvider(ServiceProvider):
    """
    Provider for Redpill AI models.

    Redpill models are Phala Cloud apps. The SDK handles verification
    specially using RedpillVerifier which uses PhalaCloudVerifier internally.
    """

    def __init__(self):
        self.api_base = "https://api.redpill.ai/v1"

    def fetch_report(self, model_id: str) -> AttestationReport:
        url = f"{self.api_base}/attestation/report"
        nonce = secrets.token_hex(32)
        print(
            f"[Redpill] Fetching from {url} for model {model_id} with nonce {nonce[:8]}..."
        )

        response = requests.get(url, params={"model": model_id, "nonce": nonce})
        response.raise_for_status()
        data = response.json()

        if "intel_quote" not in data:
            raise Exception("Redpill report missing intel_quote")

        nvidia_payload = data.get("nvidia_payload")
        if isinstance(nvidia_payload, str):
            try:
                nvidia_payload = json.loads(nvidia_payload)
            except:
                pass

        # Include model_id in raw data for verifier to look up app_id
        data["model_id"] = model_id

        return AttestationReport(
            provider="redpill",
            intel_quote=data["intel_quote"],
            request_nonce=nonce,
            nvidia_payload=nvidia_payload,
            raw=data,
        )

    def list_models(self) -> List[str]:
        url = f"{self.api_base}/models"
        print(f"[Redpill] Fetching models from {url}")
        response = requests.get(url)
        response.raise_for_status()
        data = response.json()

        models = data if isinstance(data, list) else data.get("data", [])
        return [m["id"] for m in models]
